sirepo/auth_role_moderation.py

```diff
@@ -0,0 +1,105 @@
+        except (dns.resolver.NoAnswer, dns.resolver.NXDOMAIN):
+            return 'No Reverse DNS Lookup'
+
+    def _send_moderation_email(info):
+        sirepo.smtp.send(
+            recipient=_cfg.moderator_email,
+            subject=f'{info.sim_type} Access Request',
+            body=pkjinja.render_resource('moderation_email', info),
+        )
+
+    req = sirepo.http_request.parse_post(require_sim_type=False)
+    d = req.req_data
+    uid = sirepo.auth.logged_in_user()
+
+    sirepo.auth_db.UserRoleInvite(
+        uid=uid,
+        role=sirepo.auth_role.for_sim_type(d.simulationType),
+        status='pending',
+    ).save()
+
+    _send_moderation_email(
+        PKDict(
+            display_name=sirepo.auth_db.UserRegistration.search_by(uid=uid).display_name,
+            email_addr=sirepo.auth.email.AuthEmailUser.search_by(uid=uid).unverified_email,
+            ip_addr=flask.request.remote_addr,
+            domain_name=_dns_reverse_lookup(flask.request.remote_addr),
+            link=_moderation_link(d.simulationType),
+            reason=d.reason,
+            role=sirepo.auth_role.for_sim_type(d.simulationType),
+            sim_type=d.simulationType,
+            uid=uid,
+            user_agent=flask.request.headers.get("User-Agent"),
+        )
+    )
+    return sirepo.http_reply.gen_json_ok()
+
+
+@sirepo.api_perm.require_user
+def api_getPendingUserRoleInvites():
+    sirepo.auth.check_user_is_adm()
+    rows = [row.__dict__ for row in sirepo.auth_db.UserRoleInvite.get_pending_rows()]
+    headers = [key for key in rows[0].keys() if key != '_sa_instance_state'] if rows else []
+    return sirepo.http_reply.gen_json({
+                'rows': rows,
+                'headers': headers,
+            })
+
+
+@sirepo.api_perm.require_user
+def api_approveOrDenyRole():
+    sirepo.auth.check_user_is_adm()
+    def _send_approve_or_deny_email(uid, approved, info):
+        sirepo.smtp.send(
+            recipient=sirepo.auth.get_module('email').AuthEmailUser.search_by(uid=uid).unverified_email,
+            subject=f'{info.role} Access Request' + ('Approved' if approved else 'Denied'),
+            body=pkjinja.render_resource('approved_role_email' if approved else 'denied_role_email', info),
+        )
+    req = sirepo.http_request.parse_post(require_sim_type=False, type=False)
+    uid = req.req_data.uid
+    role = req.req_data.role
+    approved = req.req_data.approved
+    role_name = sirepo.simulation_db.SCHEMA_COMMON.appInfo[sirepo.auth_role.sim_type(role)].shortName
+    link = flask.url_for('_dispatch_empty', _external=True) + sirepo.auth_role.sim_type(role)
+    if approved:
+        sirepo.auth_db.UserRole.add_role(uid, role)
+        sirepo.auth_db.UserRoleInvite.search_by(uid=uid, role=role).delete()
+        _send_approve_or_deny_email(uid, True, PKDict(link=link, role=role_name))
+        return sirepo.http_reply.gen_json_ok()
+    sirepo.auth_db.UserRoleInvite.set_status(uid, role, 'denied')
```

Comment:
I would try to reuse the word `approve` or `deny` everywhere so we don't have to remember even if it is not conjugated correctly.